### PR TITLE
ProcessApp: I try again

### PR DESCRIPTION
**Pull request by rossabaker**
We frequently get asked how to handle resource management in running an http4s app.  Our current answer of `ServerApp` is not very http4s-like: to nest resources, `shutdown` needs to be overridden with `attempt`ed tasks referring to `var`s of resources already acquired.

`Process` is designed for resource safety.  `bracket` exists to safely nest resources.  If we can express our application in terms of a `Process`, bracketed resources are closed in reverse order of acquisition as the `Process` completes.  

The other problem people have is keeping a server alive until the JVM receives a shutdown signal.  This is where `ProcessApp` comes in.  It's not specific to HTTP, but it's a huge convenience for people writing http4s apps.  Including it is a compromise between making http4s easy and not inventing generic functional resource management in an HTTP library.  Users who want flatter or more sophisticated resource management are invited to bring their own `ResourceT` or `Codensity` to the party.  This is the minimum that gets new users on their way without us mumbling, "uh, you might need a var for that."

Bonus: `ServerApp` is expressed in terms of this for a clean migration.

This will need to be `StreamApp` when merged up to cats.  It might find a more appropriate home in fs2.

### CODE REVIEW

File: core/src/main/scala/org/http4s/util/ProcessApp.scala [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  def main(args: List[String]): Process[Task, Unit]
+
+  private[this] val shutdownRequested =
+    async.signalOf(false)
+
+  final val requestShutdown: Task[Unit] =
+    shutdownRequested.set(true)
+
+  final def main(args: Array[String]): Unit = {
+    val halt = async.signalOf(false)
```
Comment (wedens) [inline core/src/main/scala/org/http4s/util/ProcessApp.scala]: looks like it's unused

Comment (wedens) [inline core/src/main/scala/org/http4s/util/ProcessApp.scala]: looks like it's unused

Comment (rossabaker) [inline core/src/main/scala/org/http4s/util/ProcessApp.scala]: Oops.  Yeah, I pulled it out so shutdowns could be requested outside a SIGTERM, and didn't clean up this one.

**Comment by rossabaker** [comment]
I'm going to merge this tomorrow if no objections arise.

File: core/src/main/scala/org/http4s/ProcessApp.scala [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+import org.http4s.internal.compatibility._
+
+import scalaz.{-\/, \/-}
+import scalaz.concurrent.Task
+import scalaz.stream.Process
+import scalaz.stream.Process.eval_
+import scalaz.stream.{async, wye}
+
+trait ProcessApp {
```
Comment (bfritz) [inline core/src/main/scala/org/http4s/ProcessApp.scala]: I'm confused about `org.http4s.ProcessApp` versus `org.http4s.util.ProcessApp`.  If I `git rm core/src/main/scala/org/http4s/ProcessApp.scala`, `sbt test` still passes.  If there's a reason we need the former, a comment might be worthwhile.

File: docs/src/main/tut/service.md [hunk 3]
```diff
@@ -129,35 +129,49 @@ A builder can be `run` to start the server.
...
+Go ahead and use curl (or your favorite HTTP client) to see your service in
+action:
+
+```sh
+$ curl http://localhost:8080/hello/Pete
+```
+
+## Cleaning up
+
+Our server consumes system resources. Let's clean up after ourselves by shutting
```
Comment (bfritz) [inline docs/src/main/tut/service.md]: Perhaps strike "after ourselves"

File: docs/src/main/tut/service.md [hunk 4]
```diff
@@ -129,35 +129,49 @@ A builder can be `run` to start the server.
 val server = builder.run
 ```
 
+Go ahead and use curl (or your favorite HTTP client) to see your service in
```
Comment (bfritz) [inline docs/src/main/tut/service.md]: Fine as-is.  Could also be "Use curl, or your favorite HTTP client, to see your service in action:".

File: docs/src/main/tut/service.md [hunk 5]
```diff
@@ -129,35 +129,49 @@ A builder can be `run` to start the server.
...
 
-To run a server as an `App` simply extend
-`org.http4s.server.ServerApp` and implement the `server` method. The server
-will be shutdown automatically when the program exits via a shutdown hook,
-so you don't need to clean up any resources explicitly.
+Every `ServerBuilder` has a `.serve` method that returns a
+`Process[Task, Nothing]`.  This process runs forever without emitting
+any output.  When this process is run with `.unsafePerformSync` on the
+main thread, it blocks forever, keeping the JVM (and your server)
+alive until it the JVM is killed.
```
Comment (bfritz) [inline docs/src/main/tut/service.md]: s/it the/the/

File: core/src/main/scala/org/http4s/ProcessApp.scala (hunk 2 above)
Comment (rossabaker) [inline core/src/main/scala/org/http4s/ProcessApp.scala]: Vestigial. I moved it without deleting.

**Pull request merged**